test/unit/org/apache/cassandra/metrics/ThreadPoolMetricsTest.java

```diff
@@ -0,0 +1,289 @@
+                                                                                 new ArrayBlockingQueue<>(2),
+                                                                                 new NamedThreadFactory("ThreadPoolMetricsTest-1"),
+                                                                                 "internal");
+        testMetricsWithNoBlockedThreads(executor, executor.metrics);
+    }
+
+    @Test
+    public void testJMXEnabledThreadPoolMetricsWithBlockedThread()
+    {
+        JMXEnabledThreadPoolExecutor executor = new JMXEnabledThreadPoolExecutor(2,
+                                                                                 Integer.MAX_VALUE,
+                                                                                 TimeUnit.SECONDS,
+                                                                                 new ArrayBlockingQueue<>(2),
+                                                                                 new NamedThreadFactory("ThreadPoolMetricsTest-2"),
+                                                                                 "internal");
+        testMetricsWithBlockedThreads(executor, executor.metrics);
+    }
+
+    @Test
+    public void testSEPExecutorMetrics()
+    {
+        SEPExecutor executor = (SEPExecutor) new SharedExecutorPool("ThreadPoolMetricsTest-2").newExecutor(2,
+                                                                                                           "ThreadPoolMetricsTest-3",
+                                                                                                           "internal");
+
+        testMetricsWithNoBlockedThreads(executor, executor.metrics);
+    }
+
+    public void testMetricsWithBlockedThreads(LocalAwareExecutorService threadPool, ThreadPoolMetrics metrics)
+    {
+        assertEquals(2, metrics.maxPoolSize.getValue().intValue());
+
+        BlockingTask task1 = new BlockingTask();
+        BlockingTask task2 = new BlockingTask();
+        BlockingTask task3 = new BlockingTask();
+        BlockingTask task4 = new BlockingTask();
+
+        // The ThreadPool has a size of 2 so the 2 first tasks should go into active straight away
+        threadPool.execute(task1);
+        threadPool.execute(task2);
+
+        spinAssertEquals(2, () -> metrics.activeTasks.getValue().intValue(), 1);
+
+        // There are no threads available any more the 2 next tasks should go into the queue
+        threadPool.execute(task3);
+        threadPool.execute(task4);
+
+        spinAssertEquals(2, () -> metrics.activeTasks.getValue().intValue(), 1);
+        spinAssertEquals(0L, () -> metrics.completedTasks.getValue().longValue(), 1);
+        spinAssertEquals(2, () -> metrics.pendingTasks.getValue().intValue(), 1);
+        spinAssertEquals(0L, () -> metrics.currentBlocked.getCount(), 1);
+        spinAssertEquals(0L, () -> metrics.totalBlocked.getCount(), 1);
```

Comment:
Not: These calls to `spinAssertEquals`, and the others below, can be slightly simplified to:
```suggestion
        spinAssertEquals(2, metrics.activeTasks::getValue, 1);
        spinAssertEquals(0L, metrics.completedTasks::getValue, 1);
        spinAssertEquals(2, metrics.pendingTasks::getValue, 1);
        spinAssertEquals(0L, metrics.currentBlocked::getCount, 1);
        spinAssertEquals(0L, metrics.totalBlocked::getCount, 1);
```
Also, we always use a timeout of one second, so we could consider having a shortcut:
```java
private static void spinAssertEquals(Object expected, Supplier<Object> actualSupplier)
{
    Util.spinAssertEquals(expected, actualSupplier, 1);
}
```